File: machine/ts/machine/devices/cartridge/controller.ts
import {
	DMA_REQUEST_CARTRIDGE_SLOT0_READ,
	DMA_REQUEST_CARTRIDGE_SLOT0_WRITE,
	DMA_REQUEST_CARTRIDGE_SLOT1_READ,
	DMA_REQUEST_CARTRIDGE_SLOT1_WRITE,
	IO_CART_SELECT,
	IO_CART_STATUS,
	IRQ_CARTRIDGE_SLOT0,
	IRQ_CARTRIDGE_SLOT1,
} from '../../../spec/bmsx/io';
import {
	CART_RAM_BASE,
} from '../../../spec/bmsx/memory_map';
import {
	MAPPED_BUS_CARTRIDGE_SLOT1,
	MAPPED_BUS_CARTRIDGE_SLOT_OVERRIDE,
	MAPPED_BUS_MASTER_CPU,
	type MappedBusSignals,
} from '../../memory/bus_signals';
import type {
	MappedPageBinding,
	MappedPageInvalidator,
	Memory,
} from '../../memory/memory';
import type { DmaController } from '../dma/controller';
import type { IrqController } from '../irq/controller';
import {
	CARTRIDGE_STATUS_SELECTED_SLOT1,
	CARTRIDGE_STATUS_SLOT0_PRESENT,
	CARTRIDGE_STATUS_SLOT1_PRESENT,
	type CartridgeSlotIndex,
} from '../../../spec/bmsx/cartridge';
import {
	type CartridgeByteView,
	type CartridgeControllerState,
	type CartridgeSocketMediaPair,
} from './contracts';
import { CartridgeCard } from './card';
import {
	CARTRIDGE_CARD_DREQ_READ,
	CARTRIDGE_CARD_DREQ_WRITE,
	CARTRIDGE_CARD_EFFECT_DREQ_CHANGED,
	CARTRIDGE_CARD_EFFECT_IRQ_EDGE,
} from './signals';

const CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET = 0x100000000;
const CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET = 0x200000000;
const CARTRIDGE_SLOT0_DREQ_MASK =
	(1 << DMA_REQUEST_CARTRIDGE_SLOT0_WRITE)
	| (1 << DMA_REQUEST_CARTRIDGE_SLOT0_READ);
const CARTRIDGE_SLOT1_DREQ_MASK =
	(1 << DMA_REQUEST_CARTRIDGE_SLOT1_WRITE)
	| (1 << DMA_REQUEST_CARTRIDGE_SLOT1_READ);

export class CartridgeController {
	private readonly cards: [CartridgeCard | null, CartridgeCard | null];
	private selectionWord = 0;
	private irq!: IrqController;
	private dma!: DmaController;

	public constructor(media: CartridgeSocketMediaPair) {
		this.cards = [
			media[0] === null
				? null
				: new CartridgeCard(media[0], CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET),
			media[1] === null
				? null
				: new CartridgeCard(media[1], CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET),
		];
	}

	public connect(memory: Memory, irq: IrqController, dma: DmaController): void {
		this.irq = irq;
		this.dma = dma;
		memory.mapIoRead(IO_CART_SELECT, this, CartridgeController.readSelectionThunk);
		memory.mapIoWrite(IO_CART_SELECT, this, CartridgeController.writeSelectionThunk);
		memory.mapIoRead(IO_CART_STATUS, this, CartridgeController.readStatusThunk);
	}

	public selectedSlot(busSignals: MappedBusSignals = MAPPED_BUS_MASTER_CPU): CartridgeSlotIndex {
		if ((busSignals & MAPPED_BUS_CARTRIDGE_SLOT_OVERRIDE) === 0) {
			return (this.selectionWord & 1) === 0 ? 0 : 1;
		}
		return (busSignals & MAPPED_BUS_CARTRIDGE_SLOT1) === 0 ? 0 : 1;
	}

	public installRom(slotIndex: number, rom: Uint8Array): void {
		this.cards[slotIndex]!.installRom(rom);
	}

	public attachMappedPageInvalidator(invalidator: MappedPageInvalidator): void {
		if (this.cards[0] !== null) this.cards[0].attachMappedPageInvalidator(invalidator);
		if (this.cards[1] !== null) this.cards[1].attachMappedPageInvalidator(invalidator);
	}

	public clearMappedPageWriteWatches(): void {
		if (this.cards[0] !== null) this.cards[0].clearMappedPageWriteWatches();
		if (this.cards[1] !== null) this.cards[1].clearMappedPageWriteWatches();
	}

	public bindMappedPage(address: number, busSignals: MappedBusSignals, out: MappedPageBinding): void {
		const slotIndex = this.selectedSlot(busSignals);
		const card = this.cards[slotIndex];
		if (card !== null) {
			card.bindMappedPage(address, out);
			return;
		}
		out.key = address + (slotIndex === 0
			? CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET
			: CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET);
		out.cacheable = address < CART_RAM_BASE;
		out.readBytes = null;
		out.readByteOffset = 0;
		out.writeWatches = null;
		out.writeWatchIndex = 0;
	}

	public ramByteCount(): number {
		return (this.cards[0] === null ? 0 : this.cards[0].ramByteCount())
			+ (this.cards[1] === null ? 0 : this.cards[1].ramByteCount());
	}

	public reset(): void {
		this.selectionWord = 0;
		if (this.cards[0] !== null) this.cards[0].reset();
		if (this.cards[1] !== null) this.cards[1].reset();
		this.publishDreqLines(0);
		this.publishDreqLines(1);
	}

	public captureState(): CartridgeControllerState {
		return {
			selectionWord: this.selectionWord,
			slots: [
				this.cards[0] === null ? null : this.cards[0].captureState(),
				this.cards[1] === null ? null : this.cards[1].captureState(),
			],
		};
	}

	public restoreState(state: CartridgeControllerState): void {
		if ((this.cards[0] === null) !== (state.slots[0] === null)
				|| (this.cards[1] === null) !== (state.slots[1] === null)) {
			throw new Error('Cartridge state does not match the occupied physical sockets.');
		}
		this.selectionWord = state.selectionWord >>> 0;
		if (this.cards[0] !== null) this.cards[0].restoreState(state.slots[0]!);
		if (this.cards[1] !== null) this.cards[1].restoreState(state.slots[1]!);
		this.publishDreqLines(0);
		this.publishDreqLines(1);
	}

	public readU8(address: number, busSignals: MappedBusSignals): number {
		const card = this.cards[this.selectedSlot(busSignals)];
		return card === null ? 0 : card.readU8(address);
	}

	public readU16(address: number, busSignals: MappedBusSignals): number {
		const card = this.cards[this.selectedSlot(busSignals)];
		return card === null ? 0 : card.readU16(address);
	}

	public readU32(address: number, busSignals: MappedBusSignals): number {
		const card = this.cards[this.selectedSlot(busSignals)];
		return card === null ? 0 : card.readU32(address);
	}

	public writeU8(address: number, value: number, busSignals: MappedBusSignals): void {
		const card = this.cards[this.selectedSlot(busSignals)];
		if (card !== null) card.writeU8(address, value);
	}

	public writeU16(address: number, value: number, busSignals: MappedBusSignals): void {
		const card = this.cards[this.selectedSlot(busSignals)];
		if (card !== null) card.writeU16(address, value);
	}

	public writeU32(address: number, value: number, busSignals: MappedBusSignals): void {
		const slotIndex = this.selectedSlot(busSignals);
		const card = this.cards[slotIndex];
		if (card === null) return;
		const effects = card.writeU32(address, value);
		if ((effects & CARTRIDGE_CARD_EFFECT_IRQ_EDGE) !== 0) {
			this.irq.raise(slotIndex === 0 ? IRQ_CARTRIDGE_SLOT0 : IRQ_CARTRIDGE_SLOT1);
		}
		if ((effects & CARTRIDGE_CARD_EFFECT_DREQ_CHANGED) !== 0) {
			this.publishDreqLines(slotIndex);
		}
	}

	public readBytes(address: number, out: Uint8Array, dstOffset: number, length: number): void {
		const card = this.cards[this.selectedSlot()];
		if (card === null) {
			out.fill(0, dstOffset, dstOffset + length);
			return;
		}
		card.readBytes(address, out, dstOffset, length);
	}

	public bindRomByteView(slotIndex: number, address: number, length: number, out: CartridgeByteView): boolean {
		const card = this.cards[slotIndex];
		return card === null ? false : card.bindRomByteView(address, length, out);
	}

	private publishDreqLines(slotIndex: CartridgeSlotIndex): void {
		let asserted = 0;
		const card = this.cards[slotIndex];
		const lines = card === null ? 0 : card.dreqLines();
		if (slotIndex === 0) {
			if ((lines & CARTRIDGE_CARD_DREQ_WRITE) !== 0) {
				asserted |= 1 << DMA_REQUEST_CARTRIDGE_SLOT0_WRITE;
			}
			if ((lines & CARTRIDGE_CARD_DREQ_READ) !== 0) {
				asserted |= 1 << DMA_REQUEST_CARTRIDGE_SLOT0_READ;
			}
			this.dma.setRequestLines(CARTRIDGE_SLOT0_DREQ_MASK, asserted);
			return;
		}
		if ((lines & CARTRIDGE_CARD_DREQ_WRITE) !== 0) {
			asserted |= 1 << DMA_REQUEST_CARTRIDGE_SLOT1_WRITE;
		}
		if ((lines & CARTRIDGE_CARD_DREQ_READ) !== 0) {
			asserted |= 1 << DMA_REQUEST_CARTRIDGE_SLOT1_READ;
		}
		this.dma.setRequestLines(CARTRIDGE_SLOT1_DREQ_MASK, asserted);
	}

	private static readSelectionThunk(context: CartridgeController): number {
		return context.selectionWord;
	}

	private static writeSelectionThunk(context: CartridgeController, _address: number, value: number): void {
		context.selectionWord = value;
	}

	private static readStatusThunk(context: CartridgeController): number {
		let status = context.selectedSlot() === 1 ? CARTRIDGE_STATUS_SELECTED_SLOT1 : 0;
		if (context.cards[0] !== null) status |= CARTRIDGE_STATUS_SLOT0_PRESENT;
		if (context.cards[1] !== null) status |= CARTRIDGE_STATUS_SLOT1_PRESENT;
		return status;
	}
}

File: machine/cpp/machine/devices/cartridge/controller.cpp
#include "machine/devices/cartridge/controller.h"

#include "machine/devices/cartridge/signals.h"
#include "machine/devices/dma/controller.h"
#include "machine/devices/irq/controller.h"
#include "machine/memory/memory.h"
#include "spec/bmsx/io.h"
#include "spec/bmsx/memory_map.h"

#include <cstring>

namespace bmsx {
namespace {

constexpr u32 CARTRIDGE_SLOT0_DREQ_MASK =
	(1u << DMA_REQUEST_CARTRIDGE_SLOT0_WRITE)
	| (1u << DMA_REQUEST_CARTRIDGE_SLOT0_READ);
constexpr u32 CARTRIDGE_SLOT1_DREQ_MASK =
	(1u << DMA_REQUEST_CARTRIDGE_SLOT1_WRITE)
	| (1u << DMA_REQUEST_CARTRIDGE_SLOT1_READ);
constexpr u64 CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET = 1ull << 32u;
constexpr u64 CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET = 2ull << 32u;

} // namespace

CartridgeController::CartridgeController(const CartridgeSocketMediaPair& media)
	: m_cards{
		media[0]
			? std::optional<CartridgeCard>(
				std::in_place,
				*media[0],
				CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET
			)
			: std::nullopt,
		media[1]
			? std::optional<CartridgeCard>(
				std::in_place,
				*media[1],
				CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET
			)
			: std::nullopt
	} {}

void CartridgeController::connect(Memory& memory, IrqController& irq, DmaController& dma) {
	m_irq = &irq;
	m_dma = &dma;
	memory.mapIoRead(IO_CART_SELECT, this, &CartridgeController::readSelectionThunk);
	memory.mapIoWrite(IO_CART_SELECT, this, &CartridgeController::writeSelectionThunk);
	memory.mapIoRead(IO_CART_STATUS, this, &CartridgeController::readStatusThunk);
}

void CartridgeController::installRom(u32 slotIndex, std::span<const u8> rom) {
	m_cards[slotIndex]->installRom(rom);
}

void CartridgeController::attachMappedPageInvalidator(
	MappedPageInvalidator& invalidator
) {
	if (m_cards[0]) m_cards[0]->attachMappedPageInvalidator(invalidator);
	if (m_cards[1]) m_cards[1]->attachMappedPageInvalidator(invalidator);
}

void CartridgeController::detachMappedPageInvalidator() {
	if (m_cards[0]) m_cards[0]->detachMappedPageInvalidator();
	if (m_cards[1]) m_cards[1]->detachMappedPageInvalidator();
}

void CartridgeController::clearMappedPageWriteWatches() {
	if (m_cards[0]) m_cards[0]->clearMappedPageWriteWatches();
	if (m_cards[1]) m_cards[1]->clearMappedPageWriteWatches();
}

void CartridgeController::bindMappedPage(
	u32 address,
	MappedBusSignals busSignals,
	MappedPageBinding& out
) {
	const u32 slotIndex = selectedSlot(busSignals);
	auto& card = m_cards[slotIndex];
	if (card) {
		card->bindMappedPage(address, out);
		return;
	}
	out.key = address + (slotIndex == 0u
		? CARTRIDGE_SLOT0_MAPPED_KEY_OFFSET
		: CARTRIDGE_SLOT1_MAPPED_KEY_OFFSET);
	out.cacheable = address < CART_RAM_BASE;
	out.readBytes = nullptr;
	out.writeWatch = nullptr;
}

void CartridgeController::reset() {
	m_selectionWord = 0u;
	if (m_cards[0]) m_cards[0]->reset();
	if (m_cards[1]) m_cards[1]->reset();
	publishDreqLines(0u);
	publishDreqLines(1u);
}

CartridgeControllerState CartridgeController::captureState(CartridgeControllerState storage) const {
	storage.selectionWord = m_selectionWord;
	for (size_t slot = 0; slot < m_cards.size(); ++slot) {
		if (!m_cards[slot]) continue;
		if (!storage.slots[slot]) storage.slots[slot].emplace();
		storage.slots[slot] = m_cards[slot]->captureState(std::move(*storage.slots[slot]));
	}
	return storage;
}

void CartridgeController::restoreState(const CartridgeControllerState& state) {
	for (u32 slotIndex = 0u; slotIndex < CARTRIDGE_SLOT_COUNT; ++slotIndex) {
		if (m_cards[slotIndex].has_value() != state.slots[slotIndex].has_value()) {
			throw BMSX_RUNTIME_ERROR(
				"Cartridge state does not match the occupied physical sockets."
			);
		}
	}
	m_selectionWord = state.selectionWord;
	for (u32 slotIndex = 0u; slotIndex < CARTRIDGE_SLOT_COUNT; ++slotIndex) {
		if (m_cards[slotIndex]) {
			m_cards[slotIndex]->restoreState(*state.slots[slotIndex]);
		}
	}
	publishDreqLines(0u);
	publishDreqLines(1u);
}

void CartridgeController::routeCardEffects(u32 slotIndex, u32 effects) {
	if ((effects & CARTRIDGE_CARD_EFFECT_IRQ_EDGE) != 0u) {
		m_irq->raise(slotIndex == 0u ? IRQ_CARTRIDGE_SLOT0 : IRQ_CARTRIDGE_SLOT1);
	}
	if ((effects & CARTRIDGE_CARD_EFFECT_DREQ_CHANGED) != 0u) {
		publishDreqLines(slotIndex);
	}
}

void CartridgeController::readBytes(u32 address, u8* out, size_t length) const {
	const auto& card = m_cards[selectedSlot()];
	if (card) {
		card->readBytes(address, out, length);
		return;
	}
	std::memset(out, 0, length);
}

bool CartridgeController::bindRomByteView(
	u32 slotIndex,
	u32 address,
	size_t length,
	Span<const u8>& out
) const {
	const auto& card = m_cards[slotIndex];
	return card ? card->bindRomByteView(address, length, out) : false;
}

void CartridgeController::publishDreqLines(u32 slotIndex) {
	u32 asserted = 0u;
	const auto& card = m_cards[slotIndex];
	const u32 lines = card ? card->dreqLines() : 0u;
	if (slotIndex == 0u) {
		if ((lines & CARTRIDGE_CARD_DREQ_WRITE) != 0u) {
			asserted |= 1u << DMA_REQUEST_CARTRIDGE_SLOT0_WRITE;
		}
		if ((lines & CARTRIDGE_CARD_DREQ_READ) != 0u) {
			asserted |= 1u << DMA_REQUEST_CARTRIDGE_SLOT0_READ;
		}
		m_dma->setRequestLines(CARTRIDGE_SLOT0_DREQ_MASK, asserted);
		return;
	}
	if ((lines & CARTRIDGE_CARD_DREQ_WRITE) != 0u) {
		asserted |= 1u << DMA_REQUEST_CARTRIDGE_SLOT1_WRITE;
	}
	if ((lines & CARTRIDGE_CARD_DREQ_READ) != 0u) {
		asserted |= 1u << DMA_REQUEST_CARTRIDGE_SLOT1_READ;
	}
	m_dma->setRequestLines(CARTRIDGE_SLOT1_DREQ_MASK, asserted);
}

u32 CartridgeController::readSelectionThunk(void* context, u32, MappedBusSignals) {
	return static_cast<CartridgeController*>(context)->m_selectionWord;
}

void CartridgeController::writeSelectionThunk(
	void* context,
	u32,
	u32 value,
	MappedBusSignals
) {
	static_cast<CartridgeController*>(context)->m_selectionWord = value;
}

u32 CartridgeController::readStatusThunk(void* context, u32, MappedBusSignals) {
	const CartridgeController& controller = *static_cast<CartridgeController*>(context);
	u32 status = controller.selectedSlot() == 1u ? CARTRIDGE_STATUS_SELECTED_SLOT1 : 0u;
	if (controller.m_cards[0]) status |= CARTRIDGE_STATUS_SLOT0_PRESENT;
	if (controller.m_cards[1]) status |= CARTRIDGE_STATUS_SLOT1_PRESENT;
	return status;
}

} // namespace bmsx
